### PR TITLE
Build Windows packages

### PR DESCRIPTION
**Pull request by leouieda**
Giving #5 a try using the instructions in https://www.ghostscript.com/doc/9.21/Make.htm#Microsoft_build_64

Checklist
* [x] Used a fork of the feedstock to propose changes
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/conda_smithy.html#how-to-re-render ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by leouieda** [comment]
Stuck on the following error after compilation: 

```
cmscgats.obj : error LNK2001: unresolved external symbol __imp_strspn
cmsps2.obj : error LNK2001: unresolved external symbol __imp__ctime64
.\bin\gsdll64.dll : fatal error LNK1120: 2 unresolved externals
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\link.EXE"' : return code '0x460'
 ```

It seems that the standard libraries aren't being found or linked correctly.

**Comment by leouieda** [comment]
@ocefpaf do you have any insight into what might be causing this error? Is there an environment variable with the correct paths to the visual studio libraries?

**Comment by ocefpaf** [comment]
Sorry it took me so long to find this ping. I'm not sure what the Windows problem :-/

@gillins and @isuruf can you see anything obvious we are missing?

```
   Creating library .\bin\gsdll64.lib and object .\bin\gsdll64.exp
LINK : warning LNK4098: defaultlib 'MSVCRT' conflicts with use of other libs; use /NODEFAULTLIB:library
cmscgats.obj : error LNK2001: unresolved external symbol __imp_strspn
cmsps2.obj : error LNK2001: unresolved external symbol __imp__ctime64
.\bin\gsdll64.dll : fatal error LNK1120: 2 unresolved externals
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\link.EXE"' : return code '0x460'
Stop.
```

**Comment by gillins** [comment]
Not sure why it is linking in `LIBCMT.lib`... Shouldn't be for a DLL. Will look into when I get a chance.

**Comment by isuruf** [comment]
Try removing `-MD` from CFLAGS, CXXFLAGS and LDFLAGS. the makefile is written to statically link to msvcrt runtime. Other option is to fix the makefile.

**Comment by gillins** [comment]
This patch seems to fix the build so it links to the runtime dynamically:
```
diff -r -u ghostscript-9.22.orig/base/msvccmd.mak ghostscript-9.22/base/msvccmd.mak
--- ghostscript-9.22.orig/base/msvccmd.mak	2017-10-04 18:53:14.000000000 +1000
+++ ghostscript-9.22/base/msvccmd.mak	2019-07-05 09:14:19.505126800 +1000
@@ -175,7 +175,7 @@
 !if $(DEBUGSYM)==0
 CT=
 LCT=
-CMT=/MT
+CMT=/MD
 COMPILE_WITHOUT_FRAMES=/Oy
 !else
 # Assume that DEBUGSYM != 0 implies a PROFILE build
diff -r -u ghostscript-9.22.orig/psi/msvc.mak ghostscript-9.22/psi/msvc.mak
--- ghostscript-9.22.orig/psi/msvc.mak	2017-10-04 18:53:14.000000000 +1000
+++ ghostscript-9.22/psi/msvc.mak	2019-07-05 09:22:29.878172700 +1000
@@ -1571,7 +1571,6 @@
 !ifdef METRO
 	echo kernel32.lib runtimeobject.lib rpcrt4.lib >> $(PSGEN)lib.rsp
 !else
-	echo LIBCMT.lib >> $(PSGEN)lib.rsp
 !endif
 !endif
 
@@ -1592,7 +1591,6 @@
 !ifdef METRO
 	echo kernel32.lib runtimeobject.lib rpcrt4.lib >> $(PCLGEN)pcllib.rsp
 !else
-	echo LIBCMT.lib >> $(PCLGEN)pcllib.rsp
 !endif
 !endif
 
@@ -1613,7 +1611,6 @@
 !ifdef METRO
 	echo kernel32.lib runtimeobject.lib rpcrt4.lib >> $(XPSGEN)xpslib.rsp
 !else
-	echo LIBCMT.lib >> $(XPSGEN)xpslib.rsp
 !endif
 !endif
 
@@ -1634,7 +1631,6 @@
 !ifdef METRO
 	echo kernel32.lib runtimeobject.lib rpcrt4.lib >> $(XPSGEN)gpdllib.rsp
 !else
-	echo LIBCMT.lib >> $(XPSGEN)gpdllib.rsp
 !endif
 !endif
 
```

But `bld.bat` seems to be missing an install step?

File: recipe/bld.bat [hunk 1]
```diff
@@ -0,0 +1,4 @@
+nmake -f psi\msvc.mak GSROOTDIR=%LIBRARY_PREFIX% WIN64= MSVC_VERSION=%VS_MAJOR% DEVSTUDIO=
+if errorlevel 1 exit 1
+
```
Comment (gillins) [inline recipe/bld.bat:3]: Doesn't seem to be an `install` target. Maybe copy the `bin`, `lib`, `Resourse` dirs like the installer does (https://ghostscript.com/doc/9.22/Install.htm#Install_Windows). I'd probably move the `.lib` files to `%LIBRARY_LIB%` and `.dll` and `.exe` files to `%LIBRARY_BIN%` too so it fits with normal conda behaviour....

File: recipe/bld.bat [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+move bin\*.dll %LIBRARY_BIN%
+cp bin\gswin64c.exe %LIBRARY_BIN%\gs.exe
```
Comment (ocefpaf) [inline recipe/bld.bat:5]: I'm not 100% sure my moving and copying here is correct. Specially the `gswin64c.exe` to `gs.exe`. It seems to work for me locally though.

**Comment by ocefpaf** [comment]
It is passing now :tada: 

Thanks @leouieda for the PR and @isuruf and @gillins for the awesome (as usual) Windows wisdom!

File: recipe/bld.bat (hunk 2 above)
Comment (gillins) [reply recipe/bld.bat:5]: I think you also need a `move bin\*.lib %LIBRARY_LIB%`. Just looking at the Linux package and it seems that the contents of the source `lib` dir has been copied to `share/ghostscript/9.22/lib` - not sure if we need to do this also. There appears to be more in the `bin` dir on Linux also which is a bit weird.
Do you know if there should be an `include` directory? Doesn't seem to be on Linux - maybe applications never link to the lib? I don't have any experience with this library, but we might only find what is missing once we try and use it elsewhere...

Comment (ocefpaf) [reply recipe/bld.bat:5]: > I think you also need a `move bin\*.lib %LIBRARY_LIB%`.

That is done in the line below, right?

> Just looking at the Linux package and it seems that the contents of the source `lib` dir has been copied to `share/ghostscript/9.22/lib` - not sure if we need to do this also.

If we don't have to I prefer to not do it. Adding versioned paths is always a headache.

> There appears to be more in the `bin` dir on Linux also which is a bit weird. Do you know if there should be an `include` directory? Doesn't seem to be on Linux - maybe applications never link to the lib? I don't have any experience with this library, but we might only find what is missing once we try and use it elsewhere...

To be honest I never checked this package the way I should and the only package depending on it is `GMT`. @leouieda do you need the header files? If I'm not mistaken all that GMT does is to call the binary, right?

File: recipe/bld.bat [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+move bin\*.exe %LIBRARY_BIN%
+move bin\*.lib %LIBRARY_LIB%
+move bin\*.exp %LIBRARY_LIB%
+move Resource %LIBRARY_PREFIX%\share
+REM fonts
```
Comment (ocefpaf) [inline recipe/bld.bat]: Not sure where to put the `fonts` folder. Maybe `%LIBRARY_PREFIX%\share`?

File: recipe/bld.bat (hunk 2 above)
Comment (gillins) [reply recipe/bld.bat:5]: > That is done in the line below, right?

Sorry - I missed that. Looks all good to me :smile:

File: recipe/bld.bat (hunk 3 above)
Comment (gillins) [reply recipe/bld.bat]: See https://www.ghostscript.com/doc/current/Use.htm#Finding_files. `GS_LIB_DEFAULT` is set by this line in `msvc.mak`:
```
GS_LIB_DEFAULT=$(GSROOTDIR)/Resource/Init;$(GSROOTDIR)/lib;$(GSROOTDIR)/Resource/Font;$(AROOTDIR)/fonts
```
So maybe one of those dirs (we don't set `AROOTDIR` ATM)? (I can't see how `GS_LIB_DEFAULT` gets passed into the build though, I must be missing something...)

Comment (ocefpaf) [reply recipe/bld.bat]: Maybe the `fonts` folder move it into `$(GSROOTDIR)/Resource/Font`? Does it make sense?

Comment (gillins) [reply recipe/bld.bat]: Sounds good to me. I guess we'll find to whether this works when we try and use it...

File: recipe/bld.bat (hunk 2 above)
Comment (leouieda) [reply recipe/bld.bat:5]: > @leouieda do you need the header files?

No, just the executable. GMT makes system calls to it.

File: recipe/bld.bat (hunk 3 above)
Comment (leouieda) [reply recipe/bld.bat]: So `fonts` would go into `%LIBRARY_PREFIX%\Resource\Font`?

**Comment by ocefpaf** [comment]
@leouieda if this last commit passes let's get it merged and test with GMT. We can try to fine tune it later.

**Comment by leouieda** [comment]
Thanks for all your inputs @ocefpaf @gillins @isuruf :beers:

File: recipe/bld.bat (hunk 3 above)
Comment (ocefpaf) [reply recipe/bld.bat]: Yep. Just fixed my last commit to reflect that.

**Pull request merged**